File: ruby/grains/grains.rb
# classic wheat and chessboard problem
module Grains
  def self.square(location)
    raise ArgumentError unless location.between?(1, 64)
    wheat = 1
    (location - 1).times do
      wheat *= 2
    end
    wheat
  end

  def self.total
    total = 0
    wheat = 1
    64.times do
      total += wheat
      wheat *= 2
    end
    total
  end
end

module BookKeeping
  VERSION = 1
end

File: c/grains/src/grains.c
#include "grains.h"

unsigned long square(short int square)
{
	if (square < 1 || square > 64)
		return 0;

	unsigned long grains = 1;
	for (int i = 1; i < square; i++)
		grains *= 2;

	return grains;
}

unsigned long total(void)
{
	return TOTAL;
}
